### PR TITLE
C++: Remove bad self joins in `cpp/toctou-race-condition`.

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR removes a bunch of bad self joins from the `cpp/toctou-race-condition` that I observed on `php`. For instance this one:

```ql
(94s) Tuple counts for #select#cpe#135#fff#shared/2@50efafh0 after 4.5s:
20213850 ~0%     {2} r1 = JOIN SSA::SsaDefinition::getAUse_dispred#fff WITH SSA::SsaDefinition::getAUse_dispred#fff ON FIRST 2 OUTPUT Rhs.2 'arg1', Lhs.2 'arg1'
                 return r1
```

The changes textual changes look quite scary, but it's really just simple refactorings following this "algorithm":
1. Pull stuff into individual predicates
2. Manually add some context into those predicates

I recommend enabling "Hide whitespace" when reviewing the PR. That makes the changes really easy to review.

### CODE REVIEW

**Review by geoffw0 — COMMENTED** [review]
Changes LGTM, I'd like to see DCA results or similar.

**Comment by geoffw0** [comment]
DCA shows a significant speedup for `cpp/toctou-race-condition` and no change in results.  :+1:

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql [hunk 1]
```diff
@@ -89,44 +89,66 @@ predicate referenceTo(Expr source, Expr use) {
...
+  // (morally, this should be a use-use pair, but it seems unlikely
+  // that this variable will get reused in practice)
+  exists(Expr call, Expr e, Variable v |
+    statCallWithPointer(checkPath, call, e, v) and
+    check.(VariableAccess).getTarget() = v and
+    not e.getAChild*() = check // the call that writes to the pointer is not where the pointer is checked.
+  )
+}
+
+predicate checkUse(Expr check, Expr checkPath, FunctionCall use, Expr usePath) {
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql:116]: There is an association: If `check` / `checkPath` is a check on a filename, then `use` / `usePath` is an operation on a filename. And if `check` / `checkPath` is another filename operation, then `use` / `usePath` looks like a sensitive operation on a filename. So I don't think we can split them up.

[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql:116]: Still looks like a cartesian product to me (or two smaller ones).  All we care about is that the other half exists in the correct form, we don't care which one it is.  So we could rewrite to something like:
```
predicate checkUse(Expr check, Expr checkPath) {
  // `check` looks like a check on a filename
  checksPath(check, checkPath) and
  exists(FunctionCall use, Expr usePath |
    use = filenameOperation(usePath)
  )
  or
  // another filename operation (null pointers can indicate errors)
  check = filenameOperation(checkPath) and
  exists(FunctionCall use, Expr usePath |
    use = sensitiveFilenameOperation(usePath)
  )
}
```
plus the corresponding predicate for `use` / `usePath`.  I guess maybe the optimizer is doing this already?

It also seems like a very strange thing to want.  I'd be surprised if this is behaving exactly as we want it to.

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql [hunk 2]
```diff
@@ -89,44 +89,66 @@ predicate referenceTo(Expr source, Expr use) {
...
+  or
+  // another filename operation (null pointers can indicate errors)
+  check = filenameOperation(checkPath) and
+  // `op` looks like a sensitive operation on a filename
+  use = sensitiveFilenameOperation(usePath)
+}
+
+pragma[noinline]
+Expr getACheckedPath(Expr check, SsaDefinition def, StackVariable v) {
+  checkUse(check, result, _, _) and
```
Comment (andersfugmann) [inline cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql:130]: Why the repeated call to `checkuse` here? Its already called on line 142. 'result' here would presumably be the same as `checkPath` from the caller point, so why not just pass the `checkPath` to the function and avoid the repeated call to `checkuse`? 

Same question for line 136.

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql (hunk 1 above)
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql:116]: Ah, I see what you mean. I'll include this change. 👍

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql]: Those repeated calls is what I referred to when I wrote that we "manually add context into those predicates". We want to ensure that we don't join `def.getAUse(v)` with itself before we restrict the set of tuples by joining with `checkUse`.

With that said, I think I've pushed another 173cefd7e4e252219b8ee663fdc6a2f6640bd05e that hopefully makes this better: Now all the columns are included in both of those predicates, and the select clause is just a giant join-on-first-6-columns now.

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql (hunk 1 above)
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql:132]: I've refactored the predicates in 173cefd7e4e252219b8ee663fdc6a2f6640bd05e to include the missing piece of information (which was still in the `select` clause) to associates `check` and `use`.

**Comment by MathiasVP** [comment]
173cefd7e4e252219b8ee663fdc6a2f6640bd05e fixes the review comments and makes the entire select clause a super crisp giant join now:
```ql
Tuple counts for #select#cpe#135#fff/3@3ee072o8 after 0ms:
  15 ~0%     {3} r1 = JOIN TOCTOUFilesystemRace::getACheckedPath#ffbfbf WITH TOCTOUFilesystemRace::getAUsedPath#ffffff ON FIRST 6 OUTPUT Lhs.3 'use', Lhs.4 'usePath', Lhs.0 'check'
              return r1
```

There's been a bunch of changes since I did the first DCA run. So I'll run another one to make sure I didn't destroy any performance on another project.

**Comment by MathiasVP** [comment]
DCA looks good.

**Comment by geoffw0** [comment]
Except I don't see a `cpp/toctou-race-condition` improvement on the "Query run timings, per query" this time.  Is there a more detailed list somewhere?

**Comment by MathiasVP** [comment]
> Except I don't see a `cpp/toctou-race-condition` improvement on the "Query run timings, per query" this time. Is there a more detailed list somewhere?

Oh, yeah. That's a good point. Here's how I concluded that we still get the performance benefits in the new DCA run:

In the first DCA run the query is mentioned in the `Query run timings, per query` table in `satisfied.md` because we were lucky enough to get a 5% performance improvement on the query (which is a magic threshold we've set that makes it appear in the table in `satisfied.md`). However, in the new DCA run we're only seeing a 3.5% performance improvement. You can see that by looking in the giant `any.md` file which contains _all_ of the query comparisons. That one has the following row:

| query                                             | `a` targets | `b` targets | a\_m   | b\_m   | diff     | relative   | conclusion |
| ------------------------------------------------- | ----------- | ----------- | ------ | ------ | -------- | ---------- | ---------- |
...
| cpp\/toctou-race-condition                        | 1           | 1           | 539.7  | 520.8  | -18.91   | -0.0350    | ✗          |
...

I think the missing 1.5% performance can be attributed to noise, so I think it's fair to conclude that we didn't lose any performance (especially considering the fact that we removed two cartesian products from the query).

**Comment by geoffw0** [comment]
Thanks, I didn't know about this information in `any.md` and I agree it looks fine.

**Review by geoffw0 — APPROVED** [review]
Happy with this, and I think all comments have been addressed.

**Pull request merged**